server/src/main/java/org/opensearch/indices/replication/common/ReplicationLuceneIndex.java

```diff
@@ -0,0 +1,523 @@
+        }
+        if (total == 0 && filesDetails.size() == 0) {
+            // indicates we are still in init phase
+            return 0.0f;
+        }
+        if (total == recovered) {
+            return 100.0f;
+        } else {
+            return 100.0f * recovered / total;
+        }
+    }
+
+    public synchronized int reusedFileCount() {
+        int reused = 0;
+        for (FileMetadata file : filesDetails.values()) {
+            if (file.reused()) {
+                reused++;
+            }
+        }
+        return reused;
+    }
+
+    public synchronized long reusedBytes() {
+        long reused = 0;
+        for (FileMetadata file : filesDetails.values()) {
+            if (file.reused()) {
+                reused += file.length();
+            }
+        }
+        return reused;
+    }
+
+    @Override
+    public synchronized XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        // stream size first, as it matters more and the files section can be long
+        builder.startObject(Fields.SIZE);
+        builder.humanReadableField(Fields.TOTAL_IN_BYTES, Fields.TOTAL, new ByteSizeValue(totalBytes()));
+        builder.humanReadableField(Fields.REUSED_IN_BYTES, Fields.REUSED, new ByteSizeValue(reusedBytes()));
+        builder.humanReadableField(Fields.RECOVERED_IN_BYTES, Fields.RECOVERED, new ByteSizeValue(recoveredBytes()));
+        builder.field(Fields.PERCENT, String.format(Locale.ROOT, "%1.1f%%", recoveredBytesPercent()));
+        builder.endObject();
+
+        builder.startObject(Fields.FILES);
+        builder.field(Fields.TOTAL, totalFileCount());
+        builder.field(Fields.REUSED, reusedFileCount());
+        builder.field(Fields.RECOVERED, recoveredFileCount());
+        builder.field(Fields.PERCENT, String.format(Locale.ROOT, "%1.1f%%", recoveredFilesPercent()));
+        filesDetails.toXContent(builder, params);
+        builder.endObject();
+        builder.humanReadableField(Fields.TOTAL_TIME_IN_MILLIS, Fields.TOTAL_TIME, new TimeValue(time()));
+        builder.humanReadableField(Fields.SOURCE_THROTTLE_TIME_IN_MILLIS, Fields.SOURCE_THROTTLE_TIME, sourceThrottling());
+        builder.humanReadableField(Fields.TARGET_THROTTLE_TIME_IN_MILLIS, Fields.TARGET_THROTTLE_TIME, targetThrottling());
+        return builder;
+    }
+
+    @Override
+    public synchronized String toString() {
+        try {
+            XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
+            builder.startObject();
+            toXContent(builder, EMPTY_PARAMS);
+            builder.endObject();
+            return Strings.toString(builder);
+        } catch (IOException e) {
+            return "{ \"error\" : \"" + e.getMessage() + "\"}";
+        }
+    }
+
+    public synchronized FileMetadata getFileDetails(String dest) {
+        return filesDetails.get(dest);
+    }
+
+    public static class FilesDetails implements ToXContentFragment, Writeable {
```

Comment:
Not used outside class. Let's mark as private and final since we don't extend.

```suggestion
    private static final class FilesDetails implements ToXContentFragment, Writeable {
```